### PR TITLE
Normalize Dialog theme

### PR DESCRIPTION
**Pull request by QuncCccccc**
This PR is to make preparations to make `DialogTheme` conform to Flutter's conventions for component themes:

* Added a `DialogThemeData` class which defines overrides for the defaults for `Dialog` properties.
* Added 2 `DialogTheme` constructor parameters: `DialogThemeData? data` and `Widget? child`. This is now the preferred way to configure a `DialogTheme`:
```
DialogTheme(
  data: DialogThemeData(color: xxx, elevation: xxx, ...),
  child: Dialog(...)
)
```
  These two properties are made nullable to not break existing apps which has customized `ThemeData.dialogTheme`.

* Changed the type of theme defaults from `DialogTheme` to `DialogThemeData`.

TODO:

* Fix internal failures.
* Change the type of `ThemeData.dialogTheme` from `DialogTheme` to `DialogThemeData`. This may cause breaking changes, a migration guide will be created.

Addresses the "theme normalization" sub project within https://github.com/flutter/flutter/issues/91772

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/test/material/dialog_theme_test.dart [hunk 1]
```diff
@@ -12,27 +12,39 @@ import 'package:flutter/material.dart';
 import 'package:flutter/rendering.dart';
 import 'package:flutter_test/flutter_test.dart';
 
-MaterialApp _appWithDialog(WidgetTester tester, Widget dialog, { ThemeData? theme }) {
+MaterialApp _appWithDialog(WidgetTester tester, Widget dialog, { ThemeData? theme, DialogThemeData? dialogTheme }) {
```
Comment (TahaTesser) [inline packages/flutter/test/material/dialog_theme_test.dart:15]:
```suggestion
MaterialApp _appWithDialog(
  WidgetTester tester,
  Widget dialog, {
  ThemeData? theme,
  DialogThemeData? dialogTheme,
}) {
```

This line got bit lengthy.

File: packages/flutter/test/material/dialog_theme_test.dart [hunk 2]
```diff
@@ -51,25 +63,77 @@ RenderParagraph _getTextRenderObject(WidgetTester tester, String text) {
   return tester.element<StatelessElement>(find.text(text)).renderObject! as RenderParagraph;
 }
 
+RenderParagraph _getIconRenderObject(WidgetTester tester, IconData icon) {
+  return tester.renderObject<RenderParagraph>(
+    find.descendant(of: find.byIcon(icon), matching: find.byType(RichText)),
+  );
+}
```
Comment (TahaTesser) [inline packages/flutter/test/material/dialog_theme_test.dart:70]:
```suggestion
RenderParagraph _getIconRenderObject(WidgetTester tester, IconData icon) {
  return tester.renderObject<RenderParagraph>(find.descendant(
    of: find.byIcon(icon),
    matching: find.byType(RichText)
  ));
}
```

This splits `descendant` for better readability.

File: packages/flutter/test/material/dialog_theme_test.dart [hunk 3]
```diff
@@ -51,25 +63,77 @@ RenderParagraph _getTextRenderObject(WidgetTester tester, String text) {
...
-  testWidgets('Dialog Theme implements debugFillProperties', (WidgetTester tester) async {
+  testWidgets('DialogThemeData implements debugFillProperties', (WidgetTester tester) async {
     final DiagnosticPropertiesBuilder builder = DiagnosticPropertiesBuilder();
-    const DialogTheme(
+    const DialogThemeData(
       backgroundColor: Color(0xff123456),
       elevation: 8.0,
       shadowColor: Color(0xff000001),
       surfaceTintColor: Color(0xff000002),
+      shape: BeveledRectangleBorder(borderRadius: BorderRadius.all(Radius.circular(20.5))),
```
Comment (TahaTesser) [inline packages/flutter/test/material/dialog_theme_test.dart:136]: NICE!

File: packages/flutter/test/material/dialog_theme_test.dart [hunk 4]
```diff
@@ -98,6 +163,76 @@ void main() {
...
+    final Material materialWidget = _getMaterialAlertDialog(tester);
+    expect(materialWidget.color, themeBackgroundColor);
+    expect(materialWidget.elevation, themeElevation);
+    expect(materialWidget.shadowColor, themeShadowColor);
+    expect(materialWidget.surfaceTintColor, themeSurfaceTintColor);
+    expect(materialWidget.shape, themeShape);
+    expect(materialWidget.clipBehavior, Clip.antiAlias);
+    final Offset bottomLeft = tester.getBottomLeft(
+      find.descendant(of: find.byType(Dialog), matching: find.byType(Material)),
+    );
```
Comment (TahaTesser) [inline packages/flutter/test/material/dialog_theme_test.dart:222]:
```suggestion
    final Offset bottomLeft = tester.getBottomLeft(find.descendant(
      of: find.byType(Dialog),
      matching: find.byType(Material)
    ));
```

File: packages/flutter/test/material/dialog_theme_test.dart [hunk 5]
```diff
@@ -98,6 +163,76 @@ void main() {
...
+    expect(bottomLeft.dx, 30.0); // 30 is the padding value.
+    expect(bottomLeft.dy, 570.0); // 600 - 30
+    expect(_getIconRenderObject(tester, Icons.search).text.style?.color, themeIconColor);
+    expect(_getTextRenderObject(tester, 'Title').text.style?.color, themeTitleTextStyle.color);
+    expect(_getTextRenderObject(tester, 'Content').text.style?.color, themeContentTextStyle.color);
+    final ModalBarrier modalBarrier = tester.widget(find.byType(ModalBarrier).last);
+    expect(modalBarrier.color, themeBarrierColor);
+
+    final Finder findPadding = find.ancestor(of: find.byIcon(Icons.cancel), matching: find.byType(Padding)).first;
```
Comment (TahaTesser) [inline packages/flutter/test/material/dialog_theme_test.dart:231]:
```suggestion
    final Finder findPadding = find.ancestor(
      of: find.byIcon(Icons.cancel),
      matching: find.byType(Padding)
    ).first;
```

**Review by TahaTesser — APPROVED** [review]
LGTM with nits!

File: packages/flutter/test/material/dialog_theme_test.dart (hunk 5 above)
Comment (QuncCccccc) [reply packages/flutter/test/material/dialog_theme_test.dart]: Done! Thanks!

File: packages/flutter/test/material/dialog_theme_test.dart (hunk 4 above)
Comment (QuncCccccc) [reply packages/flutter/test/material/dialog_theme_test.dart]: Updated!

File: packages/flutter/test/material/dialog_theme_test.dart (hunk 2 above)
Comment (QuncCccccc) [reply packages/flutter/test/material/dialog_theme_test.dart]: Updated!

File: packages/flutter/test/material/dialog_theme_test.dart (hunk 1 above)
Comment (QuncCccccc) [reply packages/flutter/test/material/dialog_theme_test.dart]: Updated!

**Pull request merged**